lib/active_merchant/billing/gateways/worldpay.rb

```diff
@@ -483,11 +543,64 @@ def required_status_message(raw, success_criteria)
         end
       end
 
-      def authorization_from(raw)
+      def authorization_from(action, raw, options)
+        order_id = order_id_from(raw)
+
+        case action
+        when 'store'
+          authorization_from_token_details(
+            order_id: order_id,
+            token_id: raw[:payment_token_id],
+            token_scope: 'shopper',
+            customer: options[:customer]
+          )
+        else
+          order_id
+        end
+      end
+
+      def order_id_from(raw)
         pair = raw.detect { |k, v| k.to_s =~ /_order_code$/ }
         (pair ? pair.last : nil)
       end
 
+      def authorization_from_token_details(options={})
+        [options[:order_id], options[:token_id], options[:token_scope], options[:customer]].join('|')
+      end
+
+      def order_id_from_authorization(authorization)
+        token_details_from_authorization(authorization)[:order_id]
+      end
+
+      def token_details_from_authorization(authorization)
+        order_id, token_id, token_scope, customer = authorization.split('|')
+
+        token_details = {}
+        token_details[:order_id] = order_id if order_id.present?
+        token_details[:token_id] = token_id if token_id.present?
+        token_details[:token_scope] = token_scope if token_scope.present?
+        token_details[:customer] = customer if customer.present?
+
+        token_details
+      end
+
+      def payment_details_from(payment_method)
+        payment_details = {}
+        if payment_method.is_a?(CreditCard)
```

Comment:
@curiousepic So I updated this to rely on our parsing the complex auth and checking that the token-specific field is in the resultant hash to determine whether it's a pay-by-order auth string or token auth string. As part of this I flipped the logic around to rely on detecting whether it's a credit card payment method using `is_a?` (I generally try really hard to avoid `is_a?`). One thing I'm uncertain of is if there is the potential for any other kind of payment_method we need to worry about at this stage which would fail this check but also not be a String (and thus cause the later code to break which assume either a `CreditCard` or `String` instance are being used).

Is there a common method that all card-like objects can respond to which could be used to differentiate them from `String` instances without having to resort to `is_a?` here?

Reply:
Looks good!  We do often use `respond_to?(:number)`.

Reply:
Ah, I like that better. Updated. Thanks!